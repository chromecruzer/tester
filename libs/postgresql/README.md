# @trac-postgresql

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test @trac-postgresql` to execute the unit tests via [Jest](https://jestjs.io).
