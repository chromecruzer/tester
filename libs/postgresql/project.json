{
  "name": "postgresql",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/postgresql/src",
  "projectType": "library",
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/postgresql/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/libs/postgresql"],
      "options": {
        "jestConfig": "libs/postgresql/jest.config.ts",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
