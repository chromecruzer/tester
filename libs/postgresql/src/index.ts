export * from './postgresql';
import PostgresConfigFactory, {PostgresConfig} from "./PostgresConfigFactory";
import BulkSqlWithCreate from "./BulkSqlWithCreate";
export *  from './ErrorHandler';
import ExtractIolData from "./ExtractIolData";
import PostgresClient from "./PostgresClient";
import LoadXlsxData from './LoadXlsxData';
import NotificationsServer from "./NotificationsServer";
import SearchTable from "./SearchTable";
import ImportIntoConsignments from "./ImportIntoConsignments";
import ImportIntoJumpTable from "./ImportIntoJumpTable";
import AuditStore from "./AuditStore";
import UserStore from './UserStore';

export {
  AuditStore, PostgresConfig, PostgresConfigFactory, BulkSqlWithCreate, ExtractIolData,
  PostgresClient, LoadXlsxData, NotificationsServer, SearchTable, ImportIntoConsignments, ImportIntoJumpTable,
  UserStore
};
