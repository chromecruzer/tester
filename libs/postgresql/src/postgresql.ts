import * as _ from "lodash";
import {DateTime} from "luxon";
import {PoolClient} from "pg";
import util from "util";
// eslint-disable-next-line @nrwl/nx/enforce-module-boundaries
import {
  customerContactDataMapping,
  customerDataMapping,
  dataTableNames, InputDataType,
  InputDataTypeLabel,
  InputToSqlMapping,
  iolDataMapping,
  matchCustomerContactFields,
  matchCustomerFields,
  matchIolFields,
  matchProductFields,
  matchSalesMappingFields,
  matchSalesRepFields,
  NullableString,
  productDataMapping,
  employeeRecordDm,
  salesMappingDataMapping,
  salesRepDataMapping,
  tracDateFormat, 
  UploadRecord,
  matchEmployeeFields,
  userRecordDm,
  matchUser1Fields
} from "../../datatypes/src";
import path from "path";

const dump = (obj, depth = null) => util.inspect(obj, {depth, colors: true});

export const auditTableNames = {
  notes: 'auditnotes',
  locations: 'auditlocation',
  items: 'audititem',
}



export const searchTableNamesFn = () => [
  dataTableNames.consignments,
  dataTableNames.customers,
  dataTableNames.customerContacts,
  dataTableNames.products,
  dataTableNames.salesReps];

export const sqlDate = (datum: (Date | string)) => {
  // console.log(`datum ${dump(datum)}
  // from JSDate ${dump(DateTime.fromJSDate(datum))}
  // to format ${dump(DateTime.fromJSDate(datum).toFormat(tracDateFormat))}`)
  const date = _.isDate(datum) ? datum : new Date(datum);
  return DateTime.fromJSDate(date).toFormat(tracDateFormat)
}

export function datumToSql(datum: (string | number | Date | boolean), datatype: InputDataTypeLabel) {
  if (_.isNull(datum) || _.isUndefined(datum) || datum === '') {
    return 'NULL';
  }
  let str;
  switch (datatype) {
    case 'string':
      str = String(datum);
      if (str.length === 0) {
        return null;
      }
      str = str.replace(/'/g, "''");
      str = str.replace(/\n/g, "\\n");
      return `'${str}'`;
    case 'date':
      // console.log(`date ${datum} converted ${sqlDate(datum as unknown as Date)}`)
      return `'${sqlDate(datum as Date)}'`
    case 'boolean':
      // console.log(`boolean ${datum} as ${Boolean(datum)}`)
      return Boolean(datum);
    case 'number':
      return datum;
  }
}

export function appendUuids(sql: string, uuids: string[]) {
  let skipped1st = false;
  let result = sql;
  uuids.forEach(id => {
    result += `${skipped1st ? ',' : ''}    \n'${id}'`;
    skipped1st = true;
  });
  result += ');';
  return result;
}

export const checkModified = modified => {
  // console.log(`Checking ${dump(modified)}`);
  if (modified.originalUuid !== modified.new.uuid) {
    console.error(`UUID mismatch found ${dump(modified)}`);
  }
}

export function appendBulkData(sql, bulkData: UploadRecord[],
                               columnMapping: InputToSqlMapping[],
                               createUuid = true,
                               hasUuid = false) {
  let skipped1stRow = false;
  let result = `${sql}\n`;
  bulkData.forEach((record) => {
    // console.log(`append bulk data ${dump(record)}`)
    const dataRow = [];
    if (createUuid) {
      dataRow.push('gen_random_uuid()')
    }
    if(hasUuid) {
      dataRow.push(datumToSql(record['uuid'], 'string'));
    }
    columnMapping.forEach((colMap) => {
      let datum = datumToSql(record[colMap.sqlLabel], colMap.dataType);
      if(!_.isUndefined(colMap.default) && datum === null) {
        datum = colMap.default;
      }
      dataRow.push(datum)
    });
    result += `${skipped1stRow ? ',': ''}\n(${_.join(dataRow, ', ')})`;
    skipped1stRow = true;
  });
  return result;
}

export const queryRecords = async (client: PoolClient, recordSql: string, idField: string,
                                   {
                                     uuid = null as NullableString,
                                     id = null as NullableString,
                                     ids = [] as string[],
                                   }) => {
  let sql = recordSql;
  const check = (_.isString(uuid) ? 1 : 0) + (_.isString(id) ? 1 : 0) + (_.isEmpty(ids) ? 0 : 1);
  if (check > 1) {
    throw new Error(`cannot set multiple parameters ${uuid}, ${id}, ${ids.length}`);
  }
  if (uuid) {
    sql += ` WHERE uuid='${uuid}'`;
  }
  if (id) {
    sql += ` WHERE ${idField}='${id}'`;
  }
  if (!_.isEmpty(ids)) {
    sql += ` WHERE ${idField} IN (${_.join(_.map(ids, i => `'${i}'`), ', ')})`;
  }
  sql += ';';
  // console.log(`query records sql '${sql}'`);
  return client.query(sql)
    .then(data => data.rows)
    .catch(err => {
    throw new Error(`Mistake in '${sql}' caused ${err.message}`);
  });
}

interface TransformerKeys {
  matchFields: (string | string[]);
  matchFn: (a: any, b: any) => boolean,
  mappings: InputToSqlMapping[];
  compareFieldList?: string[],
}

export const recordFromArray = (row: InputDataType[], mapping: InputToSqlMapping[]) => {
  return _.reduce(mapping, (accum, m,i) => {
    const r = row[i] || null;
    let value = _.isFunction(m.nullify) ? m.nullify(r) : r;
    if(_.has(m, 'default')) { //need to us _.has here because the default may be a boolean false
      // console.log(`checking default for ${dump(m)} with value ${value}`)
      value = value || m.default;
    }
    accum[m.sqlLabel] = value;
    return accum;
  }, {} as UploadRecord);
}
export const transformHelpers: { [id: string]: TransformerKeys } = {
  iol: {
    matchFields: 'lot',
    matchFn: matchIolFields,
    mappings: iolDataMapping,
  },
  customers: {
    matchFields: 'customer_code',
    matchFn: matchCustomerFields,
    mappings: customerDataMapping,
  },
  customer_contacts: {
    matchFields: ['full_name', 'ship_to_id'],
    matchFn: matchCustomerContactFields,
    mappings: customerContactDataMapping,
  },
  products: {
    matchFields: 'product_id',
    matchFn: matchProductFields,
    mappings: productDataMapping,
  },
  employee: {
     matchFields: 'emp_id',
     matchFn: matchEmployeeFields,          // added by shyam for testing
     mappings: employeeRecordDm,
  },

  users: {
    matchFields: 'username',
    matchFn:  matchUser1Fields,              // added by shyam for testing
    mappings: userRecordDm
  },
  salesmappings: {
    matchFields: 'customer_ship_to',
    matchFn: matchSalesMappingFields,
    mappings: salesMappingDataMapping,
    compareFieldList: ['customer_ship_to', 'postal', 'zipcode', 'primary_territory', 'primary_name', 'cos_territory',
      'cos_name', 'kae_territory', 'kae_name'],
  },
  salesreps: {
    matchFields: 'salesrep_id',
    matchFn: matchSalesRepFields,
    mappings: salesRepDataMapping,
  },
}
export const matchFieldFn = (fields, record) => {
  return _.isArray(fields) ? _.join(_.map(fields, f => record[f]), '|') : record[fields]
}
export const deepEqualFn = (c, m, compareFieldList = null) => {
  if (_.isArray(compareFieldList)) {
    return _.reduce(compareFieldList,
      (boolAccum, cf) => {
        if (!boolAccum) {
          return boolAccum;
        }
        // console.log(`comparing isequal field ${cf}, ${candidate[cf]} to ${m[cf]}`);
        return _.isEqual(c[cf], m[cf]);
      }, true);
  }
  return _.isEqual(m, c);
}
export const docxTemplateFn = (apiConfig, templateType) => path.join(apiConfig.templatesPath,
  (templateType === 'Missing' ? 'missingTemplate.docx' : 'expiredTemplate.docx'))
export const htmlTemplateFn = (apiConfig, templateType) => path.join(apiConfig.templatesPath,
  (templateType === 'Missing' ? 'missingTemplate.html' : 'expiredTemplate.html'))

