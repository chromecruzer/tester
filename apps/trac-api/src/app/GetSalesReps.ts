import {PostgresClient, PostgresConfigFactory, queryRecords, tryCatch} from "@trac/postgresql";
import {getDataFields} from "@trac/datatypes";
import _ from "lodash";
import * as util from "util";

const dump = (obj, depth = null) => util.inspect(obj, {depth, colors: true});

export default class GetSalesReps {
  private getSalesRepsSql: string;
  private getMappingsSql: string;
  constructor(private postgresConfig: PostgresConfigFactory, salesJumpTable: string, salesRepTable: string) {
    this.getSalesRepsSql = `SELECT * FROM ${postgresConfig.getSchemaName()}.${salesRepTable}`;
    this.getMappingsSql = `SELECT * FROM ${postgresConfig.getSchemaName()}.${salesJumpTable}`;
  }
  public getSalesReps(postgresClient: PostgresClient) {
    return tryCatch(async (req, res, next) => {
      const id = req.params[getDataFields.salesrepId] || null;
      const uuid = req.params[getDataFields.uuid] || null;
        const client = await postgresClient.getClient()
        const results = await queryRecords(client, this.getSalesRepsSql, 'salesrep_id',
          {uuid, id})
          .finally(async () => {
            console.log('trying to disconnect sales rep')
            return postgresClient.release(client);
          });
        res.json(results);
    })
  }
  public getSalesRepsForRole(postgresClient: PostgresClient) {
    return tryCatch(async (req, res, next) => {
      const role = req.params[getDataFields.role] || null;
        const client = await postgresClient.getClient()
        const mappings = await this.getMappingsForCustomer(client, role);
        const ids = _.uniq(_.map(mappings, m => m.salesrep_id));
        const salesreps =  await queryRecords(client, this.getSalesRepsSql, 'customer_code', {ids})
          .finally(async () => {
            console.log('trying to disconnect sales rep')
            return postgresClient.release(client);
          });
        res.json(_.map(salesreps, sr => ({...sr, role})));
    })
  }
  public getSalesRepsForCustomer(postgresClient: PostgresClient) {
    return tryCatch(async (req, res, next) => {
      // console.log(`mappings for req ${dump(req)} is ${dump(req.params)}`)
      const customerId = req.params[getDataFields.customerId] || null;
      const client = await postgresClient.getClient();
        const mappings = await this.getMappingsForCustomer(client, customerId);
        if(_.isEmpty(mappings)) {
          console.error(`mappings missing for customer ${customerId}`)
          res.json([]);
          return postgresClient.release(client);
        }
        const srMap = _.reduce(mappings, (accum, m) => {
          accum[m.salesrep_id] = m
          return accum;
        }, {});
        // console.log(`mapping salesreps to roles ${dump(srMap)}`)
        const salesreps = await queryRecords(client, this.getSalesRepsSql, 'salesrep_id',
          {ids:_.keys(srMap)})
          .finally(async () => {
          console.log('trying to disconnect sales rep')
          return postgresClient.release(client);
        });
        const results = _.map(salesreps, sr => ({...sr, role: srMap[sr.salesrep_id].role}));
        res.json(results);
    })
  }
  public getCustomersForSalesRep(postgresClient: PostgresClient) {
    return tryCatch(async (req, res, next) => {
      const id = req.params[getDataFields.salesrepId] || null;
      const uuid = req.params[getDataFields.uuid] || null;
      const client = await postgresClient.getClient();
        const mappings = await this.getMappingsForSalesRep(client, id, uuid);
        const ids = _.map(mappings, m => m.customer_id);
        const results =  await queryRecords(client, this.getSalesRepsSql, 'customer_code', {ids})
          .finally(async () => {
            console.log('trying to disconnect sales rep')
            return postgresClient.release(client);
          });
        res.json(results);
    })
  }
  private getMappingsForSalesRep(client, salesrepId = null, uuid = null) {
    let sql = this.getMappingsSql;
    if(salesrepId) {
      sql += ` WHERE salesrep_id = '${salesrepId}';`;
    }
    if(uuid) {
      sql += ` WHERE salesrep_uuid = '${uuid}';`;
    }
    return client.query(sql)
      .then(data => data.rows)
      .catch(err => {
        throw new Error(`Mistake in '${sql}' caused ${err.message}`);
      });
  }
  private getMappingsForCustomer(client, customerId) {
    let sql = this.getMappingsSql;
    sql += ` WHERE customer_id = '${customerId}';`;
    return client.query(sql)
      .then(data => data.rows)
      .catch(err => {
        throw new Error(`Mistake in '${sql}' caused ${err.message}`);
      });
  }
  private getMappingsForRole(client, role) {
    let sql = this.getMappingsSql;
    sql += ` WHERE role = '${role}';`;
    return client.query(sql)
      .then(data => data.rows)
      .catch(err => {
        throw new Error(`Mistake in '${sql}' caused ${err.message}`);
      });
  }
}
