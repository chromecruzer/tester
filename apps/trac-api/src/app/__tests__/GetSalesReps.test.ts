import {PostgresConfigFactory, queryRecords} from '@trac/postgresql';
import GetSalesReps from "../GetSalesReps";

jest.mock('@trac/postgresql');
describe('GetSalesReps', () => {
  let mockQueryRecords, mockClient, mockPostgresClient, getSalesReps;
  beforeEach(() => {
    mockQueryRecords = jest.mocked(queryRecords)
      .mockReturnValue(Promise.resolve(['data']));
    mockClient = {
      query: jest.fn().mockReturnValue(Promise.resolve({rows: ['data']})),
    }
    mockPostgresClient = {
      getClient: jest.fn().mockReturnValue(Promise.resolve(mockClient)),
      release: jest.fn()
    };
    const postgresConfigFactory = {
      getSchemaName: () => 'schemanameFromConfig'
    }  as PostgresConfigFactory;
    getSalesReps = new GetSalesReps(postgresConfigFactory, 'sales jump table',
      'sales rep table')
  });
  afterEach(() => {
    jest.resetAllMocks();
  });
  describe('getSalesReps', () => {
    it('should call query records correctly', () => {
      const actualFunc = getSalesReps.getSalesReps(mockPostgresClient);
      return actualFunc({params: {uuid: 'uuidForSalesRep'}}, {json: jest.fn()}, jest.fn()).then(() => {
        expect(mockQueryRecords.mock.calls).toMatchSnapshot();
      });
    });
  });
  describe('getSalesRepsForRole', () => {
    it('should query mappings and call query records correctly', () => {
      mockClient.query.mockReturnValue(Promise.resolve({ rows: [{salesrep_id: 'salesId1'},
          {salesrep_id: 'salesId2'}]}))
      const actualFunc = getSalesReps.getSalesRepsForRole(mockPostgresClient);
      return actualFunc({params: {role: 'sales rep role'}}, {json: jest.fn()}, jest.fn()).then(() => {
        expect(mockClient.query.mock.calls).toMatchSnapshot();
        expect(mockQueryRecords.mock.calls).toMatchSnapshot();
      });
    });
  });
  describe('getSalesRepsForCustomer', () => {
    it('should query mappings and call query records correctly', () => {
      mockClient.query.mockReturnValue(Promise.resolve({ rows: [{salesrep_id: 'salesId1'},
          {salesrep_id: 'salesId2'}]}))
      const actualFunc = getSalesReps.getSalesRepsForCustomer(mockPostgresClient);
      return actualFunc({params: {customerId: 'customer id'}}, {json: jest.fn()}, jest.fn()).then(() => {
        expect(mockClient.query.mock.calls).toMatchSnapshot();
        expect(mockQueryRecords.mock.calls).toMatchSnapshot();
      });
    });
  });
  describe('getCustomersForSalesRep', () => {
    it('should query mappings and call query records correctly', () => {
      mockClient.query.mockReturnValue(Promise.resolve({ rows: [{customer_id: 'customer1'},
          {customer_id: 'customer2'}]}))
      const actualFunc = getSalesReps.getCustomersForSalesRep(mockPostgresClient);
      return actualFunc({params: {salesrepId: 'sales rep id'}}, {json: jest.fn()}, jest.fn()).then(() => {
        expect(mockClient.query.mock.calls).toMatchSnapshot();
        expect(mockQueryRecords.mock.calls).toMatchSnapshot();
      });
    });
  });
});
