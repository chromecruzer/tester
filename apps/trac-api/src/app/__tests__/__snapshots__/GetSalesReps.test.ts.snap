// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetSalesReps getCustomersForSalesRep should query mappings and call query records correctly 1`] = `
Array [
  Array [
    "SELECT * FROM schemanameFromConfig.sales jump table WHERE salesrep_id = 'sales rep id';",
  ],
]
`;

exports[`GetSalesReps getCustomersForSalesRep should query mappings and call query records correctly 2`] = `
Array [
  Array [
    Object {
      "query": [MockFunction] {
        "calls": Array [
          Array [
            "SELECT * FROM schemanameFromConfig.sales jump table WHERE salesrep_id = 'sales rep id';",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Promise {},
          },
        ],
      },
    },
    "SELECT * FROM schemanameFromConfig.sales rep table",
    "customer_code",
    Object {
      "ids": Array [
        "customer1",
        "customer2",
      ],
    },
  ],
]
`;

exports[`GetSalesReps getSalesReps should call query records correctly 1`] = `
Array [
  Array [
    Object {
      "query": [MockFunction],
    },
    "SELECT * FROM schemanameFromConfig.sales rep table",
    "salesrep_id",
    Object {
      "id": null,
      "uuid": "uuidForSalesRep",
    },
  ],
]
`;

exports[`GetSalesReps getSalesRepsForCustomer should query mappings and call query records correctly 1`] = `
Array [
  Array [
    "SELECT * FROM schemanameFromConfig.sales jump table WHERE customer_id = 'customer id';",
  ],
]
`;

exports[`GetSalesReps getSalesRepsForCustomer should query mappings and call query records correctly 2`] = `
Array [
  Array [
    Object {
      "query": [MockFunction] {
        "calls": Array [
          Array [
            "SELECT * FROM schemanameFromConfig.sales jump table WHERE customer_id = 'customer id';",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Promise {},
          },
        ],
      },
    },
    "SELECT * FROM schemanameFromConfig.sales rep table",
    "salesrep_id",
    Object {
      "ids": Array [
        "salesId1",
        "salesId2",
      ],
    },
  ],
]
`;

exports[`GetSalesReps getSalesRepsForRole should query mappings and call query records correctly 1`] = `
Array [
  Array [
    "SELECT * FROM schemanameFromConfig.sales jump table WHERE customer_id = 'sales rep role';",
  ],
]
`;

exports[`GetSalesReps getSalesRepsForRole should query mappings and call query records correctly 2`] = `
Array [
  Array [
    Object {
      "query": [MockFunction] {
        "calls": Array [
          Array [
            "SELECT * FROM schemanameFromConfig.sales jump table WHERE customer_id = 'sales rep role';",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Promise {},
          },
        ],
      },
    },
    "SELECT * FROM schemanameFromConfig.sales rep table",
    "customer_code",
    Object {
      "ids": Array [
        "salesId1",
        "salesId2",
      ],
    },
  ],
]
`;
